common/readline.c

```diff
@@ -48,6 +48,20 @@ char *read_line(FILE *file) {
 	return string;
 }
 
+char *peek_line(FILE *file, int offset) {
+	int pos = ftell(file);
+	char *line = NULL;
+	for (int i = 0; i <= offset; i++) {
```

Comment:
I feel that `getline` from the standard library would be better for this, so you're not making so many allocations and then immediately freeing them.